Branch-SDK/src/io/branch/search/SearchBuilder.java

```diff
@@ -0,0 +1,146 @@
+package io.branch.search;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by sojanpr on 3/1/17.
+ * <p>
+ * Builder for Branch local search query. Use this builder to customise your search for contents
+ * with Branch search
+ * </p>
+ */
+public class SearchBuilder implements Parcelable {
+    private static final int DEF_MAX_APP_RESULT = 3;
+    private static final int DEF_MAX_CONTENT_RESULT = 5;
+    private final String query;
+    private final IBranchSearchEvent callback;
+    private final List<String> packageNames;
+    private int maxAppResults;
+    private int maxContentResults;
+
+    /**
+     * Builder for creating a query for get search result from Branch.
+     * @see  {@link #search()} method to execute a search
+     * @param query {@link String} a keyword to search
+     * @param callback {@link io.branch.search.SearchBuilder.IBranchSearchEvent} instance for listening the search results
+     */
+    public SearchBuilder(String query, IBranchSearchEvent callback) {
+        this.query = query;
+        this.callback = callback;
+        packageNames = new ArrayList<>();
+        maxAppResults = DEF_MAX_APP_RESULT;
+        maxContentResults = DEF_MAX_CONTENT_RESULT;
+    }
+
+    /**
+     * Set the maximum number of {@link AppResult} int the {@link BranchSearchResult}
+     * @param maxAppResult maximum number of app result
+     * @return SearchBuilder instance for  chaining
+     */
+    public SearchBuilder setMaxAppResults(int maxAppResult) {
+        this.maxAppResults = maxAppResult;
+        return this;
+    }
+
+    /**
+     * Set the maximum number of {@link ContentResult} int the {@link BranchSearchResult}
+     * @param maxContentResults maximum number of content result
+     * @return SearchBuilder instance for  chaining
+     */
+    public SearchBuilder setMaxContentResults(int maxContentResults) {
+        this.maxContentResults = maxContentResults;
+        return this;
+    }
+
+    /**
+     * Add the preferred  apps to search with. Setting this option will return only result associated with the specified app
+     * @param packageName package name for the preferred app
+     * @return SearchBuilder instance for  chaining
+     */
+    public SearchBuilder addPreferredApp(String packageName) {
```

Comment:
I don't think we have a usecase for this right now, so lets not add it just yet. We are not providing the ability to search just within specific apps, so lets add it only when we want to document / support the use case. But useful to know that we can support it out of the box